packages/wallet-adapter-react/src/WalletProvider.tsx

```diff
@@ -169,20 +169,11 @@ export const AptosWalletAdapterProvider: FC<AptosWalletProviderProps> = ({
     });
   }, [connected]);
 
-  // Handle the adapter's ready state change event
+  // Whenever the readyState of any supported wallet changes we produce a new wallets state array
+  // which in turn causes consumers of the `useWallet` hook to re-render.
+  // See https://github.com/aptos-labs/aptos-wallet-adapter/pull/129#issuecomment-1519026572 for reasoning.
   const handleReadyStateChange = (wallet: Wallet) => {
-    setWallets((prevWallets) => {
-      const index = prevWallets.findIndex(
-        (currWallet) => currWallet === wallet
-      );
-      if (index === -1) return prevWallets;
-
-      return [
-        ...prevWallets.slice(0, index),
-        wallet,
-        ...prevWallets.slice(index + 1),
-      ];
-    });
+    setWallets((wallets) => [...wallets]);
```

Comment:
I think you produce the same array and not really includes the new updated wallet coming from the event. Lets work on this change in a separate PR

Reply:
The "updated" wallet is already part of the previous array state. We never create "new" wallets, we just mutate their state.
Really what this function does it produce a new array with the same items as before, which is the same thing the previous code did. This refactor just makes this a bit more clear.

Reply:
Just for more clarity:

The previous code was looking for the "updated" wallet in the previous array here by using a reference comparison (i.e. it's looking for the same object):
```
const index = prevWallets.findIndex(
        (currWallet) => currWallet === wallet
      );
```

and then it produced a new array with the same object again at the same position:

```
 return [
        ...prevWallets.slice(0, index),
        wallet,
        ...prevWallets.slice(index + 1),
      ];
```

This basically produced new array with _identical_ items to the previous one, so we might as well just do `prev => [prev]` to achieve the same.

Reply:
I think it updates the array with the wallet object argument that comes from the fired event

Reply:
it does, but the wallet argument coming from the fired event is the same object reference as the object that is already in the previous array.
Otherwise
```
prevWallets.findIndex(
        (currWallet) => currWallet === wallet
      );
```
would never find a matching item in the previous array and always return `-1`.
```
currWallet === wallet
```
compares the object reference and only returns true if the left and right hand of the operator are the same object reference.

Reply:
Does it also do deep comparison? The new wallet argument should include the updated readyState value

Reply:
No, === does not do a deep comparison. It does a reference equality check checking whether `objA === objB` are indeed the same object.
<img width="815" alt="image" src="https://user-images.githubusercontent.com/1221897/233864956-c65ae870-586d-4396-8f59-843aba30d9f3.png">

Reply:
And WalletCore only sets/mutates the `.readyState` property on the existing Wallet object (that is already in the state array) and then re-emits that wallet. It never creates a new "Wallet" with a new `.readyState` property.
https://github.com/aptos-labs/aptos-wallet-adapter/blob/5afd0accc09fb67e736b80b7693ea49cac9c2088/packages/wallet-adapter-core/src/WalletCore.ts#L58-L61